File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.py
import sys
input = sys.stdin.readline

######################################
##### segment tree final version #####
######################################

def findSize():
    s = 0
    while 2**s < 2*N:
        s += 1
    return 2**s

def makeTree():
    tree = [0] * Size
    for i in range(N):
        tree[Size//2 + i] = data[i]
    for node in range(Size//2-1, 0, -1):
        tree[node] = tree[node*2] + tree[node*2+1]
    return tree

# [left, right) form
def subSum(node, nodeL, nodeR, targetL, targetR):
    if nodeR <= targetL or targetR <= nodeL:
        return 0
    if targetL <= nodeL and nodeR <= targetR:
        return segTree[node]
    mid = (nodeL + nodeR) // 2
    left = subSum(node*2, nodeL, mid, targetL, targetR)
    right = subSum(node*2+1, mid, nodeR, targetL, targetR)
    return left + right

# index start from 0
def update(i, num):
    node = Size//2 + i
    segTree[node] = num
    while node > 1:
        node = node // 2
        segTree[node] = segTree[node*2] + segTree[node*2+1]

# input data
N, M, K = map(int, input().split())
data = [int(input()) for _ in range(N)]

# build segment tree
Size = findSize()
segTree = makeTree()

# main process
for _ in range(M+K):
    a,b,c = map(int, input().split())
    if a == 1:
        update(b-1, c)
    else:
        print(subSum(1, 0, Size//2, b-1, c))

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.cc
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;
using ll = long long;

int N, M, K;
ll Data[1000000];
vector<ll> tree;
ll Size = 1;

void update(int idx, ll val){
	int node = idx + Size;
	tree[node] = val;
	while(node > 1){
		node /= 2;
		tree[node] = tree[node*2] + tree[node*2+1];
	}
}

ll query(int left, int right){
	ll rnt = 0;
	left += Size;
	right += Size;
	while(left <= right){
		if(left % 2 == 1){
			rnt += tree[left];
			left ++;
		}
		left /= 2;
		if(right % 2 == 0){
			rnt += tree[right];
			right --;
		}
		right /= 2;
	}
	return rnt;
}

int main(int argc, char** argv) {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cin >> N >> M >> K;
	for(int i = 0; i < N; i++){
		cin >> Data[i];
	}
	while(Size < N){
		Size <<= 1;
	}
	tree.resize(2*Size, 0);
	for(int i = 0; i < N; i++){
		tree[i+Size] = Data[i];
	}
	for(int i = Size-1; i > 0; i--){
		tree[i] = tree[i*2] + tree[i*2+1];
	}
	for(int i = 0; i < M + K; i++){
		ll a,b,c; cin>>a>>b>>c;
		if(a==1){
			update(b-1, c);
		}
		else{
			cout << query(b-1,c-1) << "\n";
		}
	}
}